agent/src/util/discovery_operator.rs

```diff
@@ -0,0 +1,1394 @@
+use super::super::INSTANCE_COUNT_METRIC;
+use super::{
+    constants::SHARED_INSTANCE_OFFLINE_GRACE_PERIOD_SECS,
+    device_plugin_builder::{DevicePluginBuilder, DevicePluginBuilderInterface},
+    device_plugin_service,
+    device_plugin_service::{
+        get_device_instance_name, InstanceConnectivityStatus, InstanceInfo, InstanceMap,
+    },
+    embedded_discovery_handlers::get_discovery_handler,
+    registration::{
+        DiscoveryHandlerConnectivityStatus, DiscoveryHandlerDetails, RegisteredDiscoveryHandlerMap,
+        DISCOVERY_HANDLER_OFFLINE_GRACE_PERIOD_SECS, EMBEDDED_DISCOVERY_HANDLER_ENDPOINT,
+    },
+    streaming_extension::StreamingExt,
+};
+use akri_discovery_utils::discovery::v0::{
+    discovery_client::DiscoveryClient, Device, DiscoverRequest, DiscoverResponse,
+};
+use akri_shared::{
+    akri::configuration::KubeAkriConfig,
+    k8s,
+    os::env_var::{ActualEnvVarQuery, EnvVarQuery},
+};
+use blake2::{
+    digest::{Update, VariableOutput},
+    VarBlake2b,
+};
+use log::{error, trace};
+#[cfg(test)]
+use mockall::{automock, predicate::*};
+use std::{collections::HashMap, convert::TryFrom, sync::Arc, time::Instant};
+use tokio::sync::{broadcast, mpsc};
+use tonic::{
+    transport::{Endpoint, Uri},
+    Status,
+};
+
+/// StreamType provides a wrapper around the two different types of streams returned from embedded
+/// or internal discovery handlers and ones running externally.
+pub enum StreamType {
+    Internal(mpsc::Receiver<std::result::Result<DiscoverResponse, Status>>),
+    External(tonic::Streaming<DiscoverResponse>),
+}
+
+/// A DiscoveryOperator is created for each Configuration that is applied to the cluster.
```

Comment:
For my own understanding, a DiscoveryOperator is per protocol (ie: ProtocolDiscoveryOperator)?

Reply:
It is per Configuration. Would `ConfigurationDiscoveryOperator` be better?

Reply:
Keeping it short is better. I was just confused.